scipy/stats/_bws_test.py

```diff
@@ -0,0 +1,230 @@
+import numpy as np
+from collections import namedtuple
+from ._axis_nan_policy import _axis_nan_policy_factory
+
+
+def calculate_bws_sum(ranks, nm):
+    '''Calculate the sum for the Baumgartner-Weiss-Schindler statistic'''
+    sum = 0
+    size = len(ranks)
+    factor = 1 / (size + 1)
+    for i, r in enumerate(ranks):
+        # index & ranks are 0-based, but we want 1-based
+        i1, r1 = i + 1, r + 1
+        sum += (r1 - nm / size * i1) ** 2 / (i1 * factor * (1 - i1 * factor))
+    return sum
+
+
+def bws_statistic_from_ranks(ranks, n, m):
+    '''Calculate the Baumgartner-Weiss-Schindler statistic from a set of ranks of n values from one sample.'''
+    nm = n + m
+
+    # Calculate the BWS statistic from a set of ranks of n values from one sample
+    # and m values from another sample.
+    b_x = calculate_bws_sum(ranks[:n], nm) / (m * nm)
+    b_y = calculate_bws_sum(ranks[n:], nm) / (n * nm)
+    b = (b_x + b_y) * 0.5
+
+    return b
+
+
+def shuffle(a, b):
+    '''Permutate two groups while keeping the ordering of the groups.'''
+    if not a.size:
+        yield b
+        return
+    if not b.size:
+        yield a
+        return
+    x = a[-1]
+    yield from (np.concatenate([l, [x]]) for l in shuffle(a[:-1], b))
+    x = b[-1]
+    yield from (np.concatenate([l, [x]]) for l in shuffle(a, b[:-1]))
+
+
+def bws_test_exact_from_ranks(ranks, n, m):
+    '''Exact test for the Baumgartner-Weiss-Schindler test.'''
+    b_stats = bws_statistic_from_ranks(ranks, n, m)
+
+    numlarger = 0
+    numtotal = 0
+
+    for perm in shuffle(ranks[:n], ranks[n:]):
+        perm_ranks = np.concatenate([sorted(perm[:n]), sorted(perm[n:])])
```

Comment:
You are sorting `perm` twice. Sort once then slice but then why are you concatenating again?

Reply:
nevermind you are sorting parts individually but then why slicing sort it partially.

Reply:
This is adapted from: https://math.stackexchange.com/questions/4679930/python-realization-of-permutation-of-2-or-more-groups-while-keeping-the-ordering/4679936#4679936